### PR TITLE
Add roles route to 2.0 apis

### PR DESCRIPTION
**Pull request by keedya**
depends on https://github.com/RackHD/on-core/pull/208

- Update users view and remove privilege property

@brianparry @zyoung51 @amymullins @srinia6 @dalebremner @BillyAbildgaard @geoff-reid

### CODE REVIEW

File: lib/services/account-api-service.js [hunk 1]
```diff
@@ -189,5 +197,108 @@ function accountApiServiceFactory(
...
+
+    AccountApiService.prototype.getRoleByName = function (name) {
+        var self = this;
+        return Promise.try(function () {
+            assert.string(name);
+            return waterline.roles.findOne({ role: name });
+        })
+        .then(function (roleFound) {
+            if (!roleFound) {
+                throw new Errors.NotFoundError();
```
Comment (yyscamper) [inline lib/services/account-api-service.js]: Add some message for the error.

File: lib/services/account-api-service.js [hunk 2]
```diff
@@ -59,17 +59,25 @@ function accountApiServiceFactory(
...
-                acl.addRoleParents('Administrator', ['Login','ConfigureManager','ConfigureUsers']),
-                acl.addRoleParents('Administrator', ['ConfigureComponents','ConfigureSelf']),
-                acl.addRoleParents('Operator', ['Login','ConfigureComponents','ConfigureSelf']),
-                acl.addRoleParents('ReadOnly', ['Login', 'ConfigureSelf'])
-            ]);
+                self.createRole(readOnly),
+                self.createRole(operator),
+                self.createRole(admin)
+            ]).catch(function () {
+                //mask error here
```
Comment (yyscamper) [inline lib/services/account-api-service.js]: Please fix this before removing the `WIP` flag.

File: lib/services/account-api-service.js [hunk 3]
```diff
@@ -189,5 +197,108 @@ function accountApiServiceFactory(
...
+        var acl = this.acl;
+        return Promise.try(function () {
+            assert.string(name);
+        })
+        .then(function () {
+            return waterline.roles.findOne({ role: name });
+        })
+        .then(function (entry) {
+            if (!entry) {
+                throw new Errors.NotFoundError();
```
Comment (yyscamper) [inline lib/services/account-api-service.js]: Add some message for t he threw error.

File: spec/lib/api/2.0/roles2.0-spec.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    var accountService;
+    var waterline;
+
+    before('start HTTP server', function () {
+        var self = this;
+        this.timeout(5000);
+        this.sandbox = sinon.sandbox.create();
+        return helper.startServer([])
+        .then(function () {
+            console.log('stub accountSevice');
```
Comment (yyscamper) [inline spec/lib/api/2.0/roles2.0-spec.js]: remove this before removing "WIP" flag

File: spec/lib/services/account-api-service-spec.js [hunk 5]
```diff
@@ -57,98 +70,193 @@ describe("Account API Service", function() {
...
-                expect(user.username).to.be.a('string');
+    describe('Users services', function () {
+
+        it('should list users', function () {
+            waterline.localusers.find.resolves([user1, user2]);
+            return accountService.listUsers()
+            .then(function (users) {
+                expect(waterline.localusers.find).to.have.been.called;
+                expect(users.length).to.equal(2);
+                return Promise.map(users, function (user) {
```
Comment (yyscamper) [inline spec/lib/services/account-api-service-spec.js]: Actually the `users` is already an object, you don't need to use Promise.map here.

File: spec/lib/services/account-api-service-spec.js [hunk 6]
```diff
@@ -57,98 +70,193 @@ describe("Account API Service", function() {
...
+                    expect(user.username).to.be.a('string');
+                    expect(user.password).to.be.an('undefined');
+                    expect(user.role).to.be.a('string');
+                    expect(user.privileges).to.be.instanceof(Array);
```
Comment (yyscamper) [inline spec/lib/services/account-api-service-spec.js]: For those user's property, we'd better check their content rather than only the type.

File: spec/lib/services/account-api-service-spec.js [hunk 7]
```diff
@@ -57,98 +70,193 @@ describe("Account API Service", function() {
...
-            expect(user.privileges).to.be.instanceof(Array);
-            expect(user.privileges[0]).to.equal('ConfigureUsers');
+        it('should recreate a user', function () {
+            waterline.localusers.findOne
+                .onFirstCall().resolves(user1)
+                .onSecondCall().resolves(user1);
+            waterline.localusers.create.resolves(user1);
+            return accountService.createUser(user1)
+            .then(function () {
+                expect(false);
```
Comment (yyscamper) [inline spec/lib/services/account-api-service-spec.js]: This will not work as expected

File: spec/lib/services/account-api-service-spec.js [hunk 8]
```diff
@@ -57,98 +70,193 @@ describe("Account API Service", function() {
...
+        var role1 = {
+            role: 'role1',
+            privileges: ['read']
+        };
+
+        it('should not create a new role if role already exist', function () {
+            waterline.roles.findOne.resolves(role1);
+            return accountService.createRole(role1)
+            .then(function () {
+                return new Error("expected a failure");
```
Comment (yyscamper) [inline spec/lib/services/account-api-service-spec.js]: This will not work as expected.

File: spec/lib/services/account-api-service-spec.js [hunk 9]
```diff
@@ -57,98 +70,193 @@ describe("Account API Service", function() {
...
+                expect(res[0]).to.have.property('role').that.equal(role1.role);
+                expect(res[0]).to.have.property('privileges').that.deep.equal(role1.privileges);
+            });
+        });
+
+        it('should return not found if role does not exist', function () {
+            waterline.roles.findOne.resolves();
+            return accountService.getRoleByName(role1.role)
+            .then(function (res) {
+                return new Error("expected a failure");
```
Comment (yyscamper) [inline spec/lib/services/account-api-service-spec.js]: This will not work as expected.

File: static/monorail-2.0.yaml [hunk 10]
```diff
@@ -3493,6 +3493,172 @@ paths:
...
+            type: object
+        403:
+          description: Forbidden
+          schema:
+            type: object
+        default:
+          description: Unexpected error
+          schema:
+            $ref: '#/definitions/Error'
+    post:
```
Comment (yyscamper) [inline static/monorail-2.0.yaml]: Whether the role can be duplicated? If not, change the `POST` to `PUT`

File: spec/lib/services/account-api-service-spec.js (hunk 5 above)
Comment (keedya) [inline spec/lib/services/account-api-service-spec.js]: @yyscamper this is not part of this PR. it is existing code

File: spec/lib/services/account-api-service-spec.js (hunk 6 above)
Comment (keedya) [inline spec/lib/services/account-api-service-spec.js]: @yyscamper same here

File: spec/lib/services/account-api-service-spec.js (hunk 7 above)
Comment (keedya) [inline spec/lib/services/account-api-service-spec.js]: @yyscamper this is not part of my PR

File: lib/services/account-api-service.js (hunk 2 above)
Comment (keedya) [inline lib/services/account-api-service.js]: @yyscamper this is done on purpose to mask the error. This function needs to run on startup. This part will create the essential roles (Administrator, ReadOnly and operator) if they don't exist

File: spec/lib/services/account-api-service-spec.js (hunk 7 above)
Comment (yyscamper) [inline spec/lib/services/account-api-service-spec.js]: @keedya : the existing code is not always correct. If we found some minor bug and can fix it with little effort, I strongly recommend to do this. Otherwise, we may forget it, and it's not worth to open a PR even an issue to track this.

File: spec/lib/services/account-api-service-spec.js (hunk 5 above)
Comment (yyscamper) [inline spec/lib/services/account-api-service-spec.js]: Just feel free to change existing code if it is not reasonable.

File: lib/services/account-api-service.js (hunk 2 above)
Comment (yyscamper) [inline lib/services/account-api-service.js]: If it is purposely, that's OK.

File: lib/services/account-api-service.js [hunk 11]
```diff
@@ -189,5 +199,107 @@ function accountApiServiceFactory(
...
+                role: newrole.role,
+                privileges: newrole.privileges
+            });
+        })
+        .then(function (entry) {
+            return acl.addRoleParents(entry.role, entry.privileges);
+        })
+        .then(function () {
+            return self.getRoleByName(newrole.role);
+        });
```
Comment (brianparry) [inline lib/services/account-api-service.js]: @keedya We may want to add a `.catch` here so we respond with a 409 in the event of a duplicate key error.  Otherwise, I think we'll return a 500 if two REST clients attempt to add a role concurrently.  In fact, we might not need the initial `findOne` query, we might be able to rely on the error thrown when a Mongo index violation occurs.

Maybe something like:
```
.catch(function(err) {
    if (err.message.match(/E11000 duplicate key error index:/) {
        var err = new Errors.BaseError('role already exists');
        err.status = 409;
        throw err;
    }
    throw err;
})
```

File: lib/services/account-api-service.js (hunk 1 above)
Comment (brianparry) [inline lib/services/account-api-service.js]: We could use `waterline.roles.needByIdentifier` and remove the `!roleFound` condition altogether.  I'm fine with this either way though.

File: static/monorail-2.0.yaml [hunk 12]
```diff
@@ -3493,6 +3493,172 @@ paths:
...
+            type: object
+        403:
+          description: Forbidden
+          schema:
+            type: object
+        default:
+          description: Unexpected error
+          schema:
+            $ref: '#/definitions/Error'
+    put:
```
Comment (brianparry) [inline static/monorail-2.0.yaml]: @keedya Shouldn't this be POST?

File: static/monorail-2.0.yaml [hunk 13]
```diff
@@ -3493,6 +3493,172 @@ paths:
...
+          in: body
+          description: The role information
+          required: true
+          schema:
+            $ref: '#/definitions/role_obj'
+      tags: [ "/api/2.0" ]
+      responses:
+        200:
+          description: |
+            Successfully modified the user
```
Comment (brianparry) [inline static/monorail-2.0.yaml]: Minor: user should be role

File: static/monorail-2.0.yaml [hunk 14]
```diff
@@ -3493,6 +3493,172 @@ paths:
...
+        default:
+          description: Unexpected error
+          schema:
+            $ref: '#/definitions/Error'
+    delete:
+      x-privileges: [ 'Write' ]
+      x-authentication-type: [ 'jwt' ]
+      operationId: removeRole
+      summary: |
+        Delete a user
```
Comment (brianparry) [inline static/monorail-2.0.yaml]: Minor: user should be role

File: static/monorail-2.0.yaml [hunk 15]
```diff
@@ -3493,6 +3493,172 @@ paths:
...
+        default:
+          description: Unexpected error
+          schema:
+            $ref: '#/definitions/Error'
+    get:
+      x-privileges: [ 'Write' ]
+      x-authentication-type: [ 'jwt' ]
+      operationId: getRole
+      summary: |
+        Get a user
```
Comment (brianparry) [inline static/monorail-2.0.yaml]: Minor: user should be role

**Comment by keedya** [comment]
@RackHD/corecommitters @brianparry @dalebremner @yyscamper @srinia6 @BillyAbildgaard @geoff-reid 

Addressed @brianparry  and @yyscamper comments

For the comments outside the scope of this PR, I prefer to address them in a separate PR

**Comment by brianparry** [comment]
Thanks @keedya :+1:

**Comment by yyscamper** [comment]
:+1: With a suggestion to fix the existing code's minor bug.

**Comment by keedya** [comment]
@yyscamper I am planning for some changes in users that will require changes in the unit tests as well.
Therefore the effort will be captured in a different PR.
(Changes includes making username a unique index in mongo)

**Comment by rackhdmirror** [comment]
Build finished.
